File: src/main/connection-manager.ts
import { randomUUID } from 'node:crypto';
import type { ConnectionConfig, RedisConfig } from '../shared/types';
import { createDriver } from './drivers';
import type { ConnectionDriver } from './drivers/types';
import { createRedisDriver } from './drivers/redis-factory';
import type { RedisKeyDriver } from './drivers/redis-types';
import { secretStore } from './stores/secrets';
import { connectionStore } from './stores/store';
import { NotConnectedError } from './errors';

/**
 * 运行时连接池：connectionId → 已打开的 driver。
 *
 * 设计要点：
 * - 打开后 driver 在内存里复用，避免每次操作都建立连接。
 * - 应用退出时统一关闭。
 * - 单个 driver 抛错不影响其它。
 *
 * V1 Redis 支持：kind === 'redis' 走 RedisKeyDriver 池。
 */

class ConnectionManager {
  private sqlDrivers = new Map<string, ConnectionDriver>();
  private redisDrivers = new Map<string, RedisKeyDriver>();

  /** SQL 数据库（MySQL/Oracle）打开 */
  async open(id: string, password?: string): Promise<ConnectionDriver> {
    const existing = this.sqlDrivers.get(id);
    if (existing?.isAlive()) return existing;

    const raw = connectionStore.getRaw(id);
    if (!raw) throw new NotConnectedError();
    if (raw.kind === 'redis') {
      throw new Error('该连接为 Redis，请使用 openRedis()');
    }

    const { plain: pwd, redisPwd } = await resolvePasswords(raw, password);

    const driver = createDriver(raw, pwd);
    await driver.connect();
    this.sqlDrivers.set(id, driver);
    connectionStore.touch(id);
    // 顺手记录 redis 密码（即使本次不用），方便将来 edit 后 openRedis
    if (redisPwd !== undefined) {
      // 用一个 weakmap 不太合适，这里就只缓存到下次从 store 重新拿
    }
    return driver;
  }

  /** Redis 数据库打开 —— 返回独立 driver 池 */
  async openRedis(id: string, password?: string): Promise<RedisKeyDriver> {
    const existing = this.redisDrivers.get(id);
    if (existing?.isAlive()) return existing;

    const raw = connectionStore.getRaw(id);
    if (!raw) throw new NotConnectedError();
    if (raw.kind !== 'redis') {
      throw new Error('该连接不是 Redis，请使用 open()');
    }

    const { redisPwd } = await resolvePasswords(raw, password);

    const driver = createRedisDriver(raw, redisPwd ?? '');
    await driver.connect();
    this.redisDrivers.set(id, driver);
    connectionStore.touch(id);
    return driver;
  }

  /** 同时拿到一个连接的最新 Redis 密码密文（用于 IPC handler） */
  getRedisPasswordCipher(id: string): string | undefined {
    return connectionStore.getRaw(id)?.redis?.passwordCipher;
  }

  async close(id: string): Promise<void> {
    const sql = this.sqlDrivers.get(id);
    if (sql) {
      await sql.close();
      this.sqlDrivers.delete(id);
    }
    const rds = this.redisDrivers.get(id);
    if (rds) {
      await rds.close();
      this.redisDrivers.delete(id);
    }
  }

  async closeAll(): Promise<void> {
    await Promise.all([
      ...Array.from(this.sqlDrivers.keys()).map((id) => this.close(id)),
      ...Array.from(this.redisDrivers.keys()).map((id) => this.close(id)),
    ]);
  }

  has(id: string): boolean {
    return (
      (this.sqlDrivers.has(id) && this.sqlDrivers.get(id)!.isAlive()) ||
      (this.redisDrivers.has(id) && this.redisDrivers.get(id)!.isAlive())
    );
  }

  /** 用于"无需打开连接"的纯元数据操作 */
  buildConfig(input: {
    id?: string;
    name: string;
    kind: ConnectionConfig['kind'];
    host: string;
    port: number;
    username: string;
    database?: string;
    serviceName?: string;
    sid?: string;
    tns?: string;
    charset?: string;
    timeoutMs?: number;
    redis?: RedisConfig;
    group?: string;
    color?: string;
  }): ConnectionConfig {
    const now = Date.now();
    return {
      id: input.id ?? randomUUID(),
      name: input.name.trim(),
      kind: input.kind,
      host: input.host.trim(),
      port: input.port,
      username: input.username.trim(),
      database: input.database?.trim() || undefined,
      serviceName: input.serviceName?.trim() || undefined,
      sid: input.sid?.trim() || undefined,
      tns: input.tns?.trim() || undefined,
      charset: input.charset ?? 'utf8mb4',
      timeoutMs: input.timeoutMs ?? 8000,
      redis: input.redis,
      group: input.group,
      color: input.color,
      createdAt: now,
      updatedAt: now,
    };
  }
}

/** 解出 SQL 密码 + Redis 密码（分开存放） */
async function resolvePasswords(
  raw: ConnectionConfig & { _passwordCipher?: string },
  overridePwd?: string,
): Promise<{ plain: string; redisPwd: string | undefined }> {
  let plain = overridePwd ?? '';
  if (!plain && raw._passwordCipher) {
    try {
      plain = secretStore.decrypt(raw._passwordCipher);
    } catch (_) {
      // 留空字符串
    }
  }

  let redisPwd: string | undefined;
  if (raw.redis?.passwordCipher) {
    redisPwd = secretStore.decrypt(raw.redis.passwordCipher);
  }
  return { plain, redisPwd };
}

export const connectionManager = new ConnectionManager();

File: src-tauri/src/connection_manager.rs
//! Connection manager for Eleven DB

use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::Mutex;

use crate::error::{ElevenError, Result};
use crate::types::*;
use crate::drivers::{MysqlDriver, RedisDriver};
use crate::stores::ConnectionStore;
use crate::crypto::Crypto;

pub struct ConnectionManager {
    sql_drivers: Arc<Mutex<HashMap<String, MysqlDriver>>>,
    redis_drivers: Arc<Mutex<HashMap<String, RedisDriver>>>,
    store: Arc<ConnectionStore>,
}

impl ConnectionManager {
    pub fn new(store: ConnectionStore) -> Self {
        Self {
            sql_drivers: Arc::new(Mutex::new(HashMap::new())),
            redis_drivers: Arc::new(Mutex::new(HashMap::new())),
            store: Arc::new(store),
        }
    }

    /// Open a SQL connection (MySQL/Oracle)
    pub async fn open(&self, id: &str, password: Option<String>) -> Result<MysqlDriver> {
        let mut drivers = self.sql_drivers.lock().await;
        
        // Check existing connection
        if let Some(driver) = drivers.get_mut(id) {
            if driver.is_alive() {
                return Ok(driver.clone());
            }
        }

        // Get connection config
        let record = self.store.get_raw(id)?
            .ok_or_else(|| ElevenError::ConnectionNotFound(id.to_string()))?;

        if record.config.kind == DbKind::Redis {
            return Err(ElevenError::InvalidConfig("Use open_redis for Redis connections".to_string()));
        }

        // Resolve password
        let pwd = resolve_password(&record, password)?;
        
        // Create and connect
        let mut driver = MysqlDriver::new(record.config.clone(), pwd);
        driver.connect().await?;
        
        drivers.insert(id.to_string(), driver.clone());
        self.store.touch(id)?;
        
        Ok(driver)
    }

    /// Open a Redis connection
    pub async fn open_redis(&self, id: &str, password: Option<String>) -> Result<RedisDriver> {
        let mut drivers = self.redis_drivers.lock().await;
        
        // Check existing connection
        if let Some(driver) = drivers.get_mut(id) {
            if driver.is_alive() {
                return Ok(driver.clone());
            }
        }

        // Get connection config
        let record = self.store.get_raw(id)?
            .ok_or_else(|| ElevenError::ConnectionNotFound(id.to_string()))?;

        if record.config.kind != DbKind::Redis {
            return Err(ElevenError::InvalidConfig("Use open for SQL connections".to_string()));
        }

        // Resolve Redis password
        let redis_pwd = if let Some(ref cipher) = record.redis_password_cipher {
            Some(Crypto::decrypt(cipher)?)
        } else {
            password
        };

        // Create and connect
        let mut driver = RedisDriver::new(record.config.clone(), redis_pwd);
        driver.connect().await?;
        
        drivers.insert(id.to_string(), driver.clone());
        self.store.touch(id)?;
        
        Ok(driver)
    }

    /// Close a connection
    pub async fn close(&self, id: &str) -> Result<()> {
        // Close SQL driver
        {
            let mut drivers = self.sql_drivers.lock().await;
            if let Some(mut driver) = drivers.remove(id) {
                let _ = driver.close().await;
            }
        }

        // Close Redis driver
        {
            let mut drivers = self.redis_drivers.lock().await;
            if let Some(mut driver) = drivers.remove(id) {
                let _ = driver.close().await;
            }
        }

        Ok(())
    }

    /// Close all connections
    pub async fn close_all(&self) -> Result<()> {
        // Close SQL drivers
        {
            let mut drivers = self.sql_drivers.lock().await;
            for (_, mut driver) in drivers.drain() {
                let _ = driver.close().await;
            }
        }

        // Close Redis drivers
        {
            let mut drivers = self.redis_drivers.lock().await;
            for (_, mut driver) in drivers.drain() {
                let _ = driver.close().await;
            }
        }

        Ok(())
    }

    /// Check if a connection is open
    pub async fn has(&self, id: &str) -> bool {
        let sql = self.sql_drivers.lock().await;
        if let Some(d) = sql.get(id) {
            return d.is_alive();
        }
        let redis = self.redis_drivers.lock().await;
        if let Some(d) = redis.get(id) {
            return d.is_alive();
        }
        false
    }

    /// Build a connection config
    pub fn build_config(&self, input: BuildConfigInput) -> ConnectionConfig {
        let now = chrono::Utc::now().timestamp_millis();
        ConnectionConfig {
            id: input.id.unwrap_or_else(|| uuid::Uuid::new_v4().to_string()),
            name: input.name.trim().to_string(),
            kind: input.kind,
            host: input.host.trim().to_string(),
            port: input.port,
            username: input.username.trim().to_string(),
            password_cipher: None,
            database: input.database.map(|d| d.trim().to_string()).filter(|d| !d.is_empty()),
            service_name: input.service_name.map(|s| s.trim().to_string()).filter(|s| !s.is_empty()),
            sid: input.sid.map(|s| s.trim().to_string()).filter(|s| !s.is_empty()),
            tns: input.tns.map(|t| t.trim().to_string()).filter(|t| !t.is_empty()),
            charset: input.charset.or_else(|| Some("utf8mb4".to_string())),
            timeout_ms: input.timeout_ms.or(Some(8000)),
            redis: input.redis,
            ssh: None,
            group: input.group,
            color: input.color,
            created_at: now,
            updated_at: now,
        }
    }

    /// Get store reference
    pub fn store(&self) -> &ConnectionStore {
        &self.store
    }
}

impl Clone for MysqlDriver {
    fn clone(&self) -> Self {
        Self {
            config: self.config.clone(),
            password: self.password.clone(),
            pool: None, // Connection pool cannot be cloned
        }
    }
}

impl Clone for RedisDriver {
    fn clone(&self) -> Self {
        Self {
            config: self.config.clone(),
            password: self.password.clone(),
            connection: None, // Connection cannot be cloned
            current_db: self.current_db,
        }
    }
}

fn resolve_password(record: &crate::stores::ConnectionRecord, override_pwd: Option<String>) -> Result<String> {
    if let Some(pwd) = override_pwd {
        return Ok(pwd);
    }
    
    if let Some(ref cipher) = record.password_cipher {
        return Crypto::decrypt(cipher);
    }
    
    Ok(String::new())
}
